src/backtrace/frame_pointer.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::arch::asm;
+
+use libc::c_void;
+
+#[cfg(target_os = "linux")]
+use crate::addr_validate::validate;
+
+#[derive(Clone, Debug)]
+pub struct Frame {
+    pub ip: usize,
+}
+
+extern "C" {
+    fn _Unwind_FindEnclosingFunction(pc: *mut c_void) -> *mut c_void;
+
+}
+
+impl super::Frame for Frame {
+    type S = backtrace::Symbol;
+
+    fn ip(&self) -> usize {
+        self.ip
+    }
+
+    fn resolve_symbol<F: FnMut(&Self::S)>(&self, cb: F) {
+        backtrace::resolve(self.ip as *mut c_void, cb);
+    }
+
+    fn symbol_address(&self) -> *mut libc::c_void {
+        if cfg!(target_os = "macos") || cfg!(target_os = "ios") {
+            self.ip as *mut c_void
+        } else {
+            unsafe { _Unwind_FindEnclosingFunction(self.ip as *mut c_void) }
+        }
+    }
+}
+
+pub struct Trace {}
+impl super::Trace for Trace {
+    type Frame = Frame;
+
+    fn trace<F: FnMut(&Self::Frame) -> bool>(ucontext: *mut libc::c_void, mut cb: F) {
+        let ucontext: *mut libc::ucontext_t = ucontext as *mut libc::ucontext_t;
+
+        #[cfg(all(target_arch = "x86_64", target_os = "linux"))]
+        let frame_pointer =
+            unsafe { (*ucontext).uc_mcontext.gregs[libc::REG_RBP as usize] as usize };
+
+        #[cfg(all(target_arch = "x86_64", target_os = "macos"))]
+        let frame_pointer = unsafe {
+            let mcontext = (*ucontext).uc_mcontext;
+            if mcontext.is_null() {
+                0
+            } else {
+                (*mcontext).__ss.__rbp as usize
+            }
+        };
+
+        #[cfg(all(target_arch = "aarch64", target_os = "linux"))]
+        let frame_pointer = unsafe { (*ucontext).uc_mcontext.regs[29] as usize };
+
+        // TODO: support arm64 on macos
+
+        let mut frame_pointer = frame_pointer as *mut FramePointerLayout;
+
+        // Initialize the `last_frame_pointer` with current frame pointer
+        // in case that the first frame doesn't have a frame pointer, and cause
+        // panic
```

Comment:
In what case will `the first frame doesn't have a frame pointer` appear but `RBP/x29` register is still available? Does this mean that even if the application does not enable frame-pointer, the `RBP/x29` must be correct in the signal handler?

Reply:
It's a redundant attempt to get the stack range. I should delete them.